lib/iolog/log_writer.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iolog
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"time"
+)
+
+const (
+	// MaxEntrySizeBytes is the maximum allowable size of a log entry
+	// before it will be broken into multiple entries
+	MaxEntrySizeBytes = 16 * 1024
+)
+
+// LogWriter tags log entries with descriptive metadata
+type LogWriter struct {
+	w io.Writer
+}
+
+// NewLogWriter wraps an io.WriteCloser in a LogWriter
+func NewLogWriter(w io.Writer) *LogWriter {
+	return &LogWriter{w: w}
+}
+
+// Write scans the supplied buffer, breaking off indvidual entries
+// and writing them to the underlying Writer, flusing any leftover bytes
+func (lw *LogWriter) Write(p []byte) (n int, err error) {
+	start := 0
+	for {
+		i := 0
```

Comment:
not that important but move this out of the loop not to allocate new var. in every iteration;

```
i := 0
for {
    i = 0
....
```

Same goes for end, entry and n assuming Write will be busy